### PR TITLE
TTNN Sharded tensor support

### PR DESCRIPTION
**Pull request by ntarafdar**
This adds sharded tensor support at the ttnn level.

This also simplifies the API by integrating shard_spec into memory_config. This unifies the API such that we don't need to explicitly include memory_config. 

This also pulls in the changes in this PR: https://github.com/tenstorrent-metal/tt-metal/pull/4520
(refactor shard spec underneath hood to have the buffer level include the tensor level shard spec)

### CODE REVIEW

**Comment by tarafdarTT** [comment]
@TT-billteng @DrJessop , this cherry-picks the other PR you guys reviewed

File: tt_metal/impl/buffers/buffer.hpp [hunk 1]
```diff
@@ -42,49 +42,68 @@ enum class ShardOrientation {
...
             const std::array<uint32_t, 2> & page_shape,
             const std::array<uint32_t, 2> & tensor2d_shape
-            ): ShardSpec(shard_spec)
+            ): tensor_shard_spec(shard_spec)
             {
                 this->page_shape = page_shape;
-                this-> tensor2d_size = tensor2d_shape;
+                this-> tensor2d_shape = tensor2d_shape;
             }
+    CoreRangeSet grid() {
```
Comment (TT-billteng) [inline tt_metal/impl/buffers/buffer.hpp:89]: can you mark these member functions as `const`?

File: tt_metal/impl/dispatch/command_queue.cpp [hunk 2]
```diff
@@ -393,21 +352,82 @@ const DeviceCommand EnqueueReadBufferCommand::assemble_device_command(uint32_t d
...
     command.set_num_pages(num_pages);
 
-    if (is_sharded(this->buffer.buffer_layout())) {
-        uint32_t num_cores = this->buffer.num_cores();
-        auto core_bank_indices = this->buffer.core_bank_indices();
-        command.set_sharded_buffer_num_cores(num_cores);
+    return command;
+}
+
+const DeviceCommand EnqueueReadBufferCommand::assemble_device_command_sharded(uint32_t dst_address) {
```
Comment (TT-billteng) [inline tt_metal/impl/dispatch/command_queue.cpp:358]: would it make sense to create a new class `EnqueueReadShardedBufferCommand` for shard specific code ?

File: tests/tt_eager/python_api_testing/unit_testing/test_sharded.py [hunk 3]
```diff
@@ -528,11 +528,12 @@ def test_sharded_program_cache(device, use_program_cache, function_level_default
 
 
 @pytest.mark.parametrize("in0_sharded", [True, False], ids=["in0_sharded", "in0_unsharded"])
-@pytest.mark.parametrize("out_sharded", [True, False], ids=["out_sharded", "out_unsharded"])
+# @pytest.mark.parametrize("out_sharded", [True, False], ids=["out_sharded", "out_unsharded"])
+@pytest.mark.parametrize("out_sharded", [False], ids=["out_unsharded"])
 @pytest.mark.parametrize("M", [1600])
 @pytest.mark.parametrize("N", [1024])
-@pytest.mark.parametrize("activations_dtype", [ttl.tensor.DataType.BFLOAT16, ttl.tensor.DataType.BFLOAT8_B])
-@pytest.mark.parametrize("weights_dtype", [ttl.tensor.DataType.BFLOAT16, ttl.tensor.DataType.BFLOAT8_B])
```
Comment (mywoodstock) [inline tests/tt_eager/python_api_testing/unit_testing/test_sharded.py:535]: why remove bfp8_b testing?

File: tests/tt_eager/python_api_testing/unit_testing/test_sharded.py [hunk 4]
```diff
@@ -528,11 +528,12 @@ def test_sharded_program_cache(device, use_program_cache, function_level_default
 
 
 @pytest.mark.parametrize("in0_sharded", [True, False], ids=["in0_sharded", "in0_unsharded"])
-@pytest.mark.parametrize("out_sharded", [True, False], ids=["out_sharded", "out_unsharded"])
+# @pytest.mark.parametrize("out_sharded", [True, False], ids=["out_sharded", "out_unsharded"])
```
Comment (mywoodstock) [inline tests/tt_eager/python_api_testing/unit_testing/test_sharded.py:531]: why remove testing of `out_sharded = true`?

File: tests/tt_eager/python_api_testing/unit_testing/test_sharded.py [hunk 5]
```diff
@@ -563,7 +564,7 @@ def test_sharded_matmul_2d(
 
     in0_t = torch2tt_tensor(in0, device, tt_memory_config=interleaved_mem_config, tt_dtype=activations_dtype)
     in1_t = torch2tt_tensor(in1, device, tt_memory_config=interleaved_mem_config, tt_dtype=weights_dtype)
-    bias_t = pad_by_zero(bias, device, tt_memory_config=interleaved_mem_config, tt_dtype=weights_dtype)[0]
+    # bias_t = pad_by_zero(bias, device, tt_memory_config=interleaved_mem_config, tt_dtype=weights_dtype)[0]
```
Comment (mywoodstock) [inline tests/tt_eager/python_api_testing/unit_testing/test_sharded.py:567]: why remove bias?

File: tt_eager/tt_dnn/op_library/operation.hpp [hunk 6]
```diff
@@ -218,7 +218,7 @@ struct HostOperation final {
 };
 
 struct DeviceOperation final {
-    using storage_t = std::array<std::byte, 768>;
+    using storage_t = std::array<std::byte, 816>;
```
Comment (mywoodstock) [inline tt_eager/tt_dnn/op_library/operation.hpp:221]: what is this mysterious number? can you add a comment about it?

File: ttnn/tensor.py [hunk 7]
```diff
@@ -245,6 +297,60 @@ def impl(ttl_tensor, layout):
...
+        >>> device_id = 0
+        >>> device = ttnn.open(device_id)
+        >>> tensor = ttnn.to_device(ttnn.from_torch(torch.randn((10, 64, 32), dtype=torch.bfloat16)), device)
+        >>> tensor = ttnn.to_mem_config(tensor, memory_config)
+    """
+
+    ttl_tensor = tensor.value
+    # to_sharded path
+    if memory_config.is_sharded():
+        if ttl_tensor.is_sharded():
```
Comment (mywoodstock) [inline ttnn/tensor.py:322]: Can the mem config be changing the shard type? like input is height sharded, and want to change to block sharded, etc.?

File: tt_metal/impl/dispatch/command_queue.cpp [hunk 8]
```diff
@@ -323,55 +323,12 @@ EnqueueReadBufferCommand::EnqueueReadBufferCommand(
...
-            (uint32_t)this->buffer.buffer_type(),
-            uint32_t(BufferType::SYSTEM_MEMORY),
-            this->src_page_index,
-            dst_page_index);
-    }
 
+const DeviceCommand EnqueueReadBufferCommand::assemble_device_command_postamble(const DeviceCommand & input_cmd,
+                                                    uint32_t padded_page_size,
+                                                    uint32_t num_pages
+                                                    ){
```
Comment (DrJessop) [inline tt_metal/impl/dispatch/command_queue.cpp:330]: not a fan of the tabbing + no space between ( and {

File: tt_metal/impl/dispatch/command_queue.cpp (hunk 2 above)
Comment (DrJessop) [reply tt_metal/impl/dispatch/command_queue.cpp:358]: I agree. I think EnqueueReadBuffer/EnqueueWriteBuffer can behind the scenes call the relevant API.

File: tt_metal/impl/dispatch/command_queue.cpp [hunk 9]
```diff
@@ -393,21 +352,82 @@ const DeviceCommand EnqueueReadBufferCommand::assemble_device_command(uint32_t d
...
     this->read_buffer_addr = this->manager.get_completion_queue_read_ptr(this->command_queue_channel);
 
-    const auto cmd = this->assemble_device_command(this->read_buffer_addr);
+    DeviceCommand cmd;
+    if(is_sharded(this->buffer.buffer_layout())) {
+        cmd = this->assemble_device_command_sharded(this->read_buffer_addr);
+    }
+    else {
+        cmd = this->assemble_device_command(this->read_buffer_addr);
+    }
```
Comment (DrJessop) [inline tt_metal/impl/dispatch/command_queue.cpp:430]: as @TT-billteng suggested, I think we should make new commands for sharding, and have enqueue_read_buffer create the correct command.

File: tt_metal/impl/dispatch/command_queue.hpp [hunk 10]
```diff
@@ -93,6 +97,7 @@ class EnqueueReadBufferCommand : public Command {
         std::optional<uint32_t> pages_to_read = std::nullopt);
 
     const DeviceCommand assemble_device_command(uint32_t dst);
+    const DeviceCommand assemble_device_command_sharded(uint32_t dst);
```
Comment (DrJessop) [inline tt_metal/impl/dispatch/command_queue.hpp:100]: If you do what Bill suggested, we won't need an assemble_device_command_sharded. It'll just be the assemble_device_command for EnqueueWrite/ReadBuffer

**Review by DrJessop — CHANGES_REQUESTED** [review]
I think as @TT-billteng suggested, we should create EnqueueRead/WriteShardedBufferCommand classes, and have enqueue_read_buffer/enqueue_write_buffer figure out which of those commands to construct and process. This will get rid of the need for pre-amble/post-amble and new sharded device command APIs.

File: tests/ttnn/unit_tests/test_matmul.py [hunk 11]
```diff
@@ -287,3 +287,38 @@ def test_tutorial_matmul_with_tilized_input_in_l1_memory_and_user_specified_core
...
+    input_tensor_a = ttnn.from_torch(torch_input_tensor_a)
+    input_tensor_b = ttnn.from_torch(torch_input_tensor_b)
+
+    input_tensor_a = ttnn.to_device(input_tensor_a, device, memory_config=ttnn.L1_MEMORY_CONFIG)
+    input_tensor_b = ttnn.to_device(input_tensor_b, device, memory_config=ttnn.L1_MEMORY_CONFIG)
+
+    input_tensor_a = ttnn.to_layout(input_tensor_a, ttnn.TILE_LAYOUT)
+    input_tensor_b = ttnn.to_layout(input_tensor_b, ttnn.TILE_LAYOUT)
+
+    sharded_mem_config = ttnn.BLOCK_SHARDED_MEMORY_CONFIG(grid_size, [M // grid_size[0], K // grid_size[1]])
```
Comment (arakhmati) [inline tests/ttnn/unit_tests/test_matmul.py:316]: Does `[M // grid_size[0], K // grid_size[1]]` needed as an extra argument? Can we ever pass in any other values?

**Review by arakhmati — CHANGES_REQUESTED** [review]

File: tests/tt_eager/python_api_testing/unit_testing/test_sharded.py (hunk 4 above)
Comment (tarafdarTT) [reply tests/tt_eager/python_api_testing/unit_testing/test_sharded.py]: My mistake, didn't mean to push that.

File: tests/tt_eager/python_api_testing/unit_testing/test_sharded.py (hunk 5 above)
Comment (tarafdarTT) [reply tests/tt_eager/python_api_testing/unit_testing/test_sharded.py]: My mistake, didn't mean to push that.

File: tests/tt_eager/python_api_testing/unit_testing/test_sharded.py (hunk 3 above)
Comment (tarafdarTT) [reply tests/tt_eager/python_api_testing/unit_testing/test_sharded.py]: My mistake, didn't mean to push that.

File: tt_eager/tt_dnn/op_library/operation.hpp (hunk 6 above)
Comment (tarafdarTT) [reply tt_eager/tt_dnn/op_library/operation.hpp:221]: I think it's equivalent to the size in bytes of Memory_Config. I am not sure however, @arakhmati is that the case?

File: ttnn/tensor.py (hunk 7 above)
Comment (tarafdarTT) [reply ttnn/tensor.py:322]: I can add that capability, if we are supplying a memconfig of a sharded tensor, and the tensor is already sharded.
Underneath the hood we can do sharded_to_interleaved and interleaved_to_sharded

File: ttnn/tensor.py [hunk 12]
```diff
@@ -82,6 +80,60 @@ def is_contiguous(self: "Shape") -> bool:
...
+    halo: bool = False,
+):
+    grid_coord = ttl.tensor.CoreCoord(grid[1], grid[0])
+    shard_grid = ttl.tensor.CoreRangeSet({ttl.tensor.CoreRange(ttl.tensor.CoreCoord(0, 0), grid_coord)})
+    shard_spec = ttl.tensor.ShardSpec(shard_grid, shard_shape, shard_orientation, halo)
+    mem_config = MemoryConfig(tensor_memory_layout, BufferType.L1, shard_spec)
+    return mem_config
+
+
+@decorate_operation()
```
Comment (arakhmati) [inline ttnn/tensor.py:108]: Why is decorator `decorate_operation()` used here?

File: tt_eager/tt_dnn/op_library/operation.hpp (hunk 6 above)
Comment (arakhmati) [reply tt_eager/tt_dnn/op_library/operation.hpp:221]: DeviceOperation uses type erasure to store the underlying operation struct. And the storage must be as big as the biggest underlying operation struct.

**Review by eyonland — CHANGES_REQUESTED** [review]
We need to include documentation to the matmul itself about the sharding configuration once we figure out how we want the api to work here. 

Also, is one test giving us adequate coverage here?

**Comment by tarafdarTT** [comment]
> We need to include documentation to the matmul itself about the sharding configuration once we figure out how we want the api to work here.
> 
> Also, is one test giving us adequate coverage here?

I added another test to cover another sharding method.
TBH I am not totally familiar with the matmul implementation underneath the hood (I reverse engineered the parameters to feed in by looking at the asserts). @TT-BrianLiu could you look at adding some documentation regarding sharding for the matmul?

File: tt_metal/impl/dispatch/command_queue.cpp [hunk 13]
```diff
@@ -322,54 +322,70 @@ EnqueueReadBufferCommand::EnqueueReadBufferCommand(
...
+}
+
+const DeviceCommand EnqueueReadInterleavedBufferCommand::create_buffer_transfer_instruction(uint32_t dst_address, uint32_t padded_page_size, uint32_t num_pages) {
+    DeviceCommand command;
+    TT_ASSERT(not is_sharded(this->buffer.buffer_layout()));
+
+    uint32_t buffer_address = this->buffer.address();
+    uint32_t dst_page_index = 0;
+
+    command.add_buffer_transfer_instruction(
```
Comment (DrJessop) [inline tt_metal/impl/dispatch/command_queue.cpp:371]: See my command about naming this function in device_command.cpp

**Review by DrJessop — APPROVED** [review]
Approving so as not to block you, but I believe addressing my new comments should be quite quick and would improve the code readability.

**Review by arakhmati — APPROVED** [review]

File: tests/ttnn/unit_tests/test_matmul.py [hunk 14]
```diff
@@ -287,3 +287,78 @@ def test_tutorial_matmul_with_tilized_input_in_l1_memory_and_user_specified_core
...
+    output = ttnn.from_device(output)
+    output = ttnn.to_torch(output)
+
+    assert_with_pcc(torch_output_tensor, output, pcc=0.999)
+
+
+@skip_for_wormhole_b0()
+def test_height_sharded_matmul(device):
+    torch.manual_seed(0)
+    grid_size = (8, 12)
```
Comment (mywoodstock) [inline tests/ttnn/unit_tests/test_matmul.py:330]: shouldn't this be `(12, 8)`? We use `x, y` for grids

**Comment by TT-BrianLiu** [comment]
Why do we need `ttnn.create_height_sharded_memory_config`, `ttnn.create_width_sharded_memory_config`, and `ttnn.create_block_sharded_memory_config`? All we are hiding is just one `ShardingStrategy` variable. I think we just need `create_sharded_memory_config`. 

Also, I don't think we should have a default for shard orientation. User should be conscious of this since matmul has requirements on shard orientation. (Unless we are okay with errors being thrown back at the user).

**Comment by TT-BrianLiu** [comment]
Can we also directly do:
```
input_tensor_a = ttnn.from_torch(torch_input_tensor_a, layout, sharded_memory_config)
```
instead of:
```
input_tensor_a = ttnn.from_torch(torch_input_tensor_a)
input_tensor_a = ttnn.to_device(input_tensor_a, device, memory_config=ttnn.DRAM_MEMORY_CONFIG)
input_tensor_a = ttnn.to_layout(input_tensor_a, ttnn.TILE_LAYOUT)
input_tensor_a = ttnn.to_memory_config(input_tensor_a, sharded_mem_config)
```

**Comment by tarafdarTT** [comment]
> Can we also directly do:
> 
> ```
> input_tensor_a = ttnn.from_torch(torch_input_tensor_a, layout, sharded_memory_config)
> ```
> 
> instead of:
> 
> ```
> input_tensor_a = ttnn.from_torch(torch_input_tensor_a)
> input_tensor_a = ttnn.to_device(input_tensor_a, device, memory_config=ttnn.DRAM_MEMORY_CONFIG)
> input_tensor_a = ttnn.to_layout(input_tensor_a, ttnn.TILE_LAYOUT)
> input_tensor_a = ttnn.to_memory_config(input_tensor_a, sharded_mem_config)
> ```

Currently we also do support

```
input_tensor_a = ttnn.from_torch(torch_input_tensor_a)
input_tensor_a = ttnn.to_device(input_tensor_a, device, sharded_mem_config)
```

Underneath the hood though it does something slightly different.
The example you posted (going to device and then to_memory_config) copies to device as an interleaved buffer (`to_device`), and then `to_memory_config` does the `interleaved_to_sharded`.

The example I posted below, moves directly to device as a sharded buffer. Our infra supports that if we use the sharded_mem_config instead. I wrote the matmul examples using interleaved_to_sharded to be consistent with examples we already have in tt_sharded.

I think we should leave both paths as users may want to either move to interleaved first and then convert or straight to sharded. 

Whether or not we want to combine `to_device` and `from_torch` in one command, that I think is a separate discussion. Maybe outside the scope of this PR.

**Comment by tarafdarTT** [comment]
hey @eyonland just waiting for your approval. the scope of this PR, is to add sharding support at the TTNN level, I have added a couple matmul examples but I don't think its super thorough. @TT-BrianLiu is working on a subsequent PR working on this PR that will have more support for matmul.

**Review by boris-drazic — APPROVED** [review]

File: tests/ttnn/unit_tests/test_matmul.py (hunk 14 above)
Comment (tarafdarTT) [reply tests/ttnn/unit_tests/test_matmul.py:356]: @mywoodstock at the TTNN level y,x are swapped when inputting core_grid. I decided to leave my sharding grid to be consistent with the rest of ttnn

**Comment by arakhmati** [comment]
> Can we also directly do:
> 
> ```
> input_tensor_a = ttnn.from_torch(torch_input_tensor_a, layout, sharded_memory_config)
> ```
> 
> instead of:
> 
> ```
> input_tensor_a = ttnn.from_torch(torch_input_tensor_a)
> input_tensor_a = ttnn.to_device(input_tensor_a, device, memory_config=ttnn.DRAM_MEMORY_CONFIG)
> input_tensor_a = ttnn.to_layout(input_tensor_a, ttnn.TILE_LAYOUT)
> input_tensor_a = ttnn.to_memory_config(input_tensor_a, sharded_mem_config)
> ```



@TT-BrianLiu 

Yes, we can do that now:
```
input_tensor_a = ttnn.from_torch(torch_input_tensor_a, layout=layout, device=device, memory_config=sharded_memory_config)
```

But as @tarafdarTT

**Comment by davorchap** [comment]
> > Can we also directly do:
> > ```
> > input_tensor_a = ttnn.from_torch(torch_input_tensor_a, layout, sharded_memory_config)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > instead of:
> > ```
> > input_tensor_a = ttnn.from_torch(torch_input_tensor_a)
> > input_tensor_a = ttnn.to_device(input_tensor_a, device, memory_config=ttnn.DRAM_MEMORY_CONFIG)
> > input_tensor_a = ttnn.to_layout(input_tensor_a, ttnn.TILE_LAYOUT)
> > input_tensor_a = ttnn.to_memory_config(input_tensor_a, sharded_mem_config)
> > ```
> 
> @TT-BrianLiu
> 
> Yes, we can do that now:
> 
> ```
> input_tensor_a = ttnn.from_torch(torch_input_tensor_a, layout=layout, device=device, memory_config=sharded_memory_config)
> ```
> 
> But as @tarafdarTT mentioned, we want to keep different paths

having both paths of doing is great

**Review by davorchap — APPROVED** [review]

**Review by farbabi — APPROVED** [review]

File: docs/source/ttnn/tensor.rst [hunk 15]
```diff
@@ -138,6 +138,50 @@ Storage
...
+***************
+
+Sharding refers to a tensor that is split across a distributed memory space, with that distribution ideally being abstracted away.
+Currently we support L1 sharding, which refers to the distribution of a tensor across the L1 memory space of different cores.
+
+Sharded tensors are represented in two dimensions, we compress a higher rank tensor into two-dims by compressing all upper dims in dim 0 and keeping the last dimension the same.
+For example a [1,2,3,4] tensor will be represented as a [6,4] 2D tensor before sharding.
+Data in a :class:`ttnn.Tensor` is organized in tiles, and these tiles are typically in row-major order.
+The size of the tile depends on the :class:`ttnn.Layout` and :class:`ttnn.Shape` of the tensor.
+Refer to the section about ``Layout`` to learn more.
```
Comment (eyonland) [inline docs/source/ttnn/tensor.rst:151]: Replace 
```
``Layout``
```
with
```
:ref:`Layout<ttnn.Layout>`
```

File: docs/source/ttnn/tensor.rst [hunk 16]
```diff
@@ -138,6 +138,50 @@ Storage
...
+
+A sharded tensor's tiles are organized in sharded order on the device which we will highlight with an example.
+A tensor is only sharded on device and on the host the tiles remain in row-major order.
+Below is an example of a tensor on the host, with each page labelled.
```
Comment (eyonland) [inline docs/source/ttnn/tensor.rst:155]: Can we create a link to page (like we have for ttnn.Layout) that explains what a page is?

**Review by eyonland — CHANGES_REQUESTED** [review]
Great work!    Love this documentation.   Just two minor changes and it looks good to go!

File: docs/source/ttnn/tensor.rst (hunk 16 above)
Comment (tarafdarTT) [reply docs/source/ttnn/tensor.rst]: hey eon, sorry I meant a tile, which we already explain above.

File: docs/source/ttnn/tensor.rst (hunk 15 above)
Comment (tarafdarTT) [reply docs/source/ttnn/tensor.rst]: fixed that here: https://github.com/tenstorrent-metal/tt-metal/pull/4597/commits/1c8231d97e2faba7aa786ce9f7630fbd2fd325a0

**Review by arakhmati — APPROVED** [review]

**Review by eyonland — APPROVED** [review]

File: docs/source/ttnn/tensor.rst [hunk 17]
```diff
@@ -138,6 +138,57 @@ Storage
...
+Data in a :class:`ttnn.Tensor` is organized in tiles, and these tiles are typically in row-major order.
+The size of the tile depends on the :class:`ttnn.Layout` and :class:`ttnn.Shape` of the tensor.
+Refer to the section about :ref:`Layout<ttnn.Layout>` to learn more.
+
+**Data Organization on Host**
+
+A sharded tensor's tiles are organized in sharded order on the device which we will highlight with an example.
+A tensor is only sharded on device and on the host the tiles remain in row-major order.
+Below is an example of a tensor on the host, with each tile labelled.
+
```
Comment (mywoodstock) [inline docs/source/ttnn/tensor.rst:161]: The above seems to imply that sharding only works with `TILE`. I think sharding description should be independent of the layout, and then mention that both TILE and RM layouts can be used.

File: docs/source/ttnn/tensor.rst [hunk 18]
```diff
@@ -138,6 +138,57 @@ Storage
...
+Tensor Sharding
+***************
+
+Sharding refers to a tensor that is split across a distributed memory space, with that distribution ideally being abstracted away.
+Currently we support L1 sharding, which refers to the distribution of a tensor across the L1 memory space of different cores.
+
+
+
+Sharded tensors are represented in two dimensions, we compress a higher rank tensor into two-dims by compressing all upper dims in dim 0 and keeping the last dimension the same.
+For example a [1,2,3,4] tensor will be represented as a [6,4] 2D tensor before sharding.
```
[the remark this answers is not in the document]
Comment (tarafdarTT) [reply docs/source/ttnn/tensor.rst:151]: the user doesn't do either, this is for us to have a conceptual understanding of sharding so the user can understand the diagrams that come after.

File: docs/source/ttnn/tensor.rst (hunk 17 above)
Comment (tarafdarTT) [reply docs/source/ttnn/tensor.rst:161]: this is to remain consistent with the rest of the documentation where in the layout section we already defined a tile. I do not want to confuse the user more by having them differentiate between tiles and pages.

File: tt_metal/impl/dispatch/command_queue.hpp [hunk 19]
```diff
@@ -83,6 +85,10 @@ class EnqueueReadBufferCommand : public Command {
     static constexpr EnqueueCommandType type_ = EnqueueCommandType::ENQUEUE_READ_BUFFER;
     uint32_t command_queue_id;
 
+
+    virtual const DeviceCommand create_buffer_transfer_instruction(uint32_t dst_address, uint32_t padded_page_size, uint32_t num_pages) = 0;
+    friend class EnqueueReadShardedBufferCommand;
```
Comment (TT-billteng) [inline tt_metal/impl/dispatch/command_queue.hpp:90]: why is this needed?

File: tt_metal/impl/dispatch/command_queue.hpp [hunk 20]
```diff
@@ -113,6 +166,10 @@ class EnqueueWriteBufferCommand : public Command {
     uint32_t pages_to_write;
     static constexpr EnqueueCommandType type_ = EnqueueCommandType::ENQUEUE_WRITE_BUFFER;
     uint32_t command_queue_id;
+
+    virtual const DeviceCommand create_buffer_transfer_instruction(uint32_t dst_address, uint32_t padded_page_size, uint32_t num_pages) = 0;
+    friend class EnqueueWriteShardedBufferCommand;
```
Comment (TT-billteng) [inline tt_metal/impl/dispatch/command_queue.hpp:171]: why is this needed?

**Review by farbabi — APPROVED** [review]

**Review by TT-billteng — APPROVED** [review]

**Pull request merged**